### ISSUE TITLE
PDP 24: Rolling Transaction. 

### ISSUE THREAD

**Issue by shiveshr** [pravega/pravega#2045]
**Problem description**
Discuss PDP 24. 

**Problem location**

**Suggestions for an improvement**

**Comment by andreipaduroiu** [comment]
> There are two ways in which we can achieve (2) -
>  1.  "Elevate" transaction segment by renaming it to follow regular segment's naming convention. (introduce rename segment primitive in segment store)
>  2.  "Elevate" transaction segment by manipulating metadata in controller. (Note: we do not rename the segment but use txn segments original name which contains reference to parent segment).

> Since 1 will have a performance cost penalty while 2 has additional code complexity on controller, my suggestion is to choose 2.

Hmm, this may be very tricky. If all we do is a Controller metadata change, then that will have adverse impacts in the SegmentStore. The Transactions are loosely coupled with their parent segments via the name (if the parent is called "foo", its transactions will be called "foo#txn-id"). Regardless of what you do in the Controller, the SegmentStore will always treat a transaction as tied to its parent, and will **only** merge it into its parent based on the name. Therefore, if the parent is deleted, the transaction segment becomes orphaned, and if the parent is sealed, the transaction also becomes "useless" since it cannot be merged anymore.

Removing this coupling in the SegmentStore would be akin to ripping through the code in numerous components and possibly relaxing constraints that guarantee data integrity. I would not do this.

**Comment by shiveshr** [comment]
How costly is it to relax the requirement that txn segments can only be merged to parent? Can we merge them to siblings instead if parent no longer exists. Also based on an earlier discussion, orphaned transaction segments are not GC'd. So even if parent is sealed/deleted they can be accessed and if we change client to take segment id as a string instead of a number, then clients can connect to these orphaned txn-segments to read from. 

Can we introduce following flavours of txn merges:

`merge(baseSegment, segmentToMerge1, segmentToMerge2, ...)`
constraints: BaseSegment is either parent or sibling of segments to merge.

Then this api can be used in two ways:
1. merge(parentSegemnt, txnSegment1, txnSegment2, ...) // when txns are merged into parent
2. merge(txnSegemnt1, txnSegment2, txnSegment3, ...) // when txns are orphaned as parent is sealed (possibly deleted in case of truncation)

**Comment by andreipaduroiu** [comment]
The SegmentStore's API treats Transactions like any other Segments, which means you can write to them, read from them, delete them, seal them, call GetInfo, etc.

You can also call `merge(TransactionName)` which  will use the internal metadata (via the naming convention I brought up in the previous comment) to determine which Segment to merge it into.

The only reason why there exists a concept of a Transaction in the SegmentStore is because Transactions must live in the same SegmentContainer as the Segment they will eventually be merged in. Other than that, there is absolutely no reason why there should even be a concept of a "Transaction" in the SegmentStore. Any segment can be merged into any other segment, but we just enforce a constraint that the source must be one of the target's transactions. The only other place where transactions are used is when a parent segment is deleted (all its known transactions are also deleted).

Now, if we figure out a way to keep all such related segments mapped to the same Segment Container, then we can completely get rid of the concept of Transactions in the Segment Store and relax the constraint of what can be merged into what. This is not a simple change (it will take more than 2 weeks to do) simply because of the number of components that need to be touched and all the tests that need to be updated (and any new ones).

The major piece we would need to solve (in Pravega, not just in SegmentStore) is:
- Given a Segment name, generate a new Segment name that will map to the same SegmentContainer as the original one. This new Segment is either a "continuation" of the old one (in case of scaling) or a "transaction".

Once we solve that, the rest is simply a bunch of time consuming find-and-fix problem to chase away transactions from the Segment Store.

**Comment by shiveshr** [comment]
> Given a Segment name, generate a new Segment name that will map to the same SegmentContainer as the original one. This new Segment is either a "continuation" of the old one (in case of scaling) or a "transaction".

This is not very difficult. In the example i gave above, thats why i said, if we merge sibling txns into one another with new name following the original naming convention (parentSegment#uniqueTxnId) then the mapping is implicitly taken care of. 
Yes this segment will look unusual after parent segment is sealed. But since it does not matter if it is orphaned as controller has decided in its metadata that this segment now constitutes a first class segment for the stream, so thats all that matters.

**Comment by andreipaduroiu** [comment]
It's not just transactions. Your continuation segments will need to end up on the same Container, which means you can't just assign the next available id to it; you'd need to generate a name based on the original Segment's name.

A side effect of this will be a skew of Segments to Containers, which will prevent us from uniformly distributing loads across Containers.

**Comment by shiveshr** [comment]
> A side effect of this will be a skew of Segments to Containers, which will prevent us from uniformly distributing loads across Containers.
In the proposal, if the epoch against which txn was created is no longer active, then txns when merged, will be accumulated into a new epoch. 

example:
epoch = 0    segments = (S1, S2)
say we have segment S1 and S2 on different segment containers. 
Create n transactions: t1, t2, t3 ... tn (so txn segments = {S1#t1, S2#t1}, {S1#t2, S2#t2}... {S1#tn, S2#tn})

say transactions 1 to k are merged within this epoch:
`merge (S1, S1#t1, S1#t2, .. S1#tk)` ==> merges segments into S1 so segment container ownership doesnt change.
similarly, 
`merge (S2, S2#t1, S2#t2, .. S2#tk)` ==> merges segments into S2 so segment container ownership doesnt change.

epoch = 1  segments = S3, S4, S2    // S1 scaled up into S3 and S4. This means S1 is sealed. 

Now commit txn comes for remainder of transactions:
`merge(S1#tK+1, S1#tK+2 ... S1#tn)` ==> merges segments into S1#tk+1, and still segment container ownership does not change for final remaining segment `S1#tk+1`
`merge(S2#tK+1, S2#tK+2... S2#tn)` ==> merged into `S2#tk+1`   
// Note: S2 is not sealed, but since epoch has moved ahead, we will elevate S2#tk+1 as first class segment. Controller governs the fact that it will never merge S2#tk+2 into S2.

If you notice, whatever was segment distribution across containers before transactions, the same factor is maintained. 
It is correct that the load on segmentContainers which had ownership of such segments increases while remainder segment containers do not get ownership of these new segments that were created. 
But we can assume that transactions will be randomly distributed across streams and their respective segments. 
So it may all even out.

**Comment by andreipaduroiu** [comment]
Ah, so you don't want to merge `S1#tK+1, S1#tK+2 ... S1#tn` into S3 or S4? You only want to merge them into `S1#tk+1`? Is my understanding correct? If so, then it should be simpler, in that we don't need to worry about force-pinning segments to containers.

**Comment by shiveshr** [comment]
Yes, the only drawback, as you pointed out earlier, is uneven distribution of segments across segment containers.

And one more change we will require is segment Id no longer remains a `number` but becomes a string. 

The change on controller will be quite significant for this notion of segment number. 
It will touch every query done in controller where we access metadata. 
But this takes advantage of the fact that we can manipulate segments on controller in any way we want as long as we are satisfying all invariants defined for a stream.

**Comment by tkaitchuck** [comment]
@shiveshr There is no big issue on the client. It would need some changes, probably an API to query the the controller to list the segment names between two points. But not a big deal.
